server/src/main/java/org/elasticsearch/index/query/InterceptedQueryBuilderWrapper.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.query;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.TransportVersion;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+
+/**
+ * Wrapper for queries that have been intercepted using the {@link QueryRewriteInterceptor} that may need to
+ * break out of the rewrite phase.
+ * @param <T>
+ */
+public class InterceptedQueryBuilderWrapper<T extends AbstractQueryBuilder<T>> extends AbstractQueryBuilder<T> {
```

Comment:
I think this is a generally good idea. Could we just call it `AbstractQueryBuilderWrapper`? 

Additionally, could you add a note about how the query is unwrapped on serialization. Meaning, only the particular instance of the object is wrapped?

Reply:
Addressed in 1c696d024858f757e1ddfe25394ffc32853b2319